### ISSUE TITLE
Calling for Improving Robustness of FSDP-vLLM Rollout

### ISSUE THREAD

**Issue by nwiad** [volcengine/verl#48]
### Issue
I encountered a presumably deadlock issue in PPO training. The tracking run was created but the program got stuck in the initial validation. It failed to generate any sequences and hung until it was killed.

The console output is as follows:
```
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|wandb: ⭐️ View project at https://xxxxx
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|wandb: 🚀 View run at https://xxxxx
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|wandb: Tracking run with wandb version 0.13.75
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|wandb: Run data is saved locally in /opt/xxxxx
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|wandb: Run `wandb offline` to turn off syncing.
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.out|None|Using LocalLogger is deprecated. The constructor API will change 
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|WARNING:2024-12-13 19:04:43,105:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fa9b47e6590>, 'Connection to o151352.ingest.sentry.io timed out. (connect timeout=<object object at 0x7fdcec53c400>)')': /api/5288891/envelope/
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|WARNING:2024-12-13 19:04:45,665:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fa9b482a690>, 'Connection to o151352.ingest.sentry.io timed out. (connect timeout=<object object at 0x7fdcec53c400>)')': /api/5288891/envelope/
worker-24bacbc3b7179c8f29db83624033f607c7b8082317d843d5913f105d-01000000-9283.err|None|WARNING:2024-12-13 19:04:48,225:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fa9b4830410>, 'Connection to o151352.ingest.sentry.io timed out. (connect timeout=<object object at 0x7fdcec53c400>)')': /api/5288891/envelope/
worker-03d79421f92cd30ae091e6311017f263220550b08b429234e5137092-01000000-13632.err|WorkerDict|[rank7]:[E1213 19:23:29.516423745 ProcessGroupNCCL.cpp:1375] [PG 3 Rank 7] First PG on this rank that detected no heartbeat of its watchdog.
worker-03d79421f92cd30ae091e6311017f263220550b08b429234e5137092-01000000-13632.err|WorkerDict|[rank7]:[E1213 19:23:29.516514634 ProcessGroupNCCL.cpp:1413] [PG 3 Rank 7] Heartbeat monitor timed out! Process will be terminated after dumping debug info. workMetaList_.size()=0
worker-03d79421f92cd30ae091e6311017f263220550b08b429234e5137092-01000000-13632.err|WorkerDict|[rank7]:[F1213 19:33:29.517864369 ProcessGroupNCCL.cpp:1224] [PG 3 Rank 7] [PG 3 Rank 7] ProcessGroupNCCL's watchdog got stuck for 600 seconds without making progress in monitoring enqueued collectives. This typically indicates a NCCL/CUDA API hang blocking the watchdog, and could be triggered by another thread holding the GIL inside a CUDA api, or other deadlock-prone behaviors.If you suspect the watchdog is not actually stuck and a longer timeout would help, you can either increase the timeout (TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC) to a larger value or disable the heartbeat monitor (TORCH_NCCL_ENABLE_MONITORING=0).If either of aforementioned helps, feel free to file an issue to PyTorch about the short timeout or false positive abort; otherwise, please attempt to debug the hang. workMetaList_.size() = 0
```

And here is a script for reproducing the issue:
```
VAL_BS=500
ROLLOUT_TP=8
GPUS_PER_NODE=8

python3 -m verl.trainer.main_ppo \
    data.train_files=$DATA_DIR/train.parquet \
    data.val_files=$DATA_DIR/test.parquet \
    data.train_batch_size=128 \
    data.val_batch_size=$VAL_BS \
    data.max_prompt_length=1024 \
    data.max_response_length=512 \
    actor_rollout_ref.model.path=meta-llama/Llama-3.1-8B-Instruct \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.ppo_micro_batch_size=16 \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.grad_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size=16 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=$ROLLOUT_TP \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.4 \
    actor_rollout_ref.ref.log_prob_micro_batch_size=16 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    critic.optim.lr=1e-5 \
    critic.model.path=meta-llama/Llama-3.1-8B-Instruct \
    critic.model.enable_gradient_checkpointing=False \
    critic.ppo_micro_batch_size=16 \
    critic.model.fsdp_config.param_offload=False \
    critic.model.fsdp_config.grad_offload=False \
    critic.model.fsdp_config.optimizer_offload=False \
    algorithm.kl_ctrl.kl_coef=0.001 \
    trainer.critic_warmup=0 \
    trainer.logger=['console','tracking'] \
    trainer.project_name='verl_example' \
    trainer.experiment_name='llama-3.1-8b-instruct_gsm8k_ppo' \
    trainer.n_gpus_per_node=$GPUS_PER_NODE \
    trainer.nnodes=1 \
    trainer.save_freq=-1 \
    trainer.test_freq=10 \
    trainer.total_epochs=15
```

### Workaround
I managed to work around by changing `actor_rollout_ref.rollout.tensor_model_parallel_size` from 8 to 1.

### Cause of the Issue
I figure the cause for this issue is that the validation batch size (500) cannot be evenly divided by the world size (8). When I use `tensor_model_parallel_size = 8`, the program will attempt to perform `all_gather` to collect **uneven sized tensors** from 8 different ranks, which is **not supported by PyTorch < 2.5.0** (https://pytorch.org/docs/2.4/distributed.html#torch.distributed.all_gather). On the other hand, when I set `tensor_model_parallel_size = 1`, there is no need to perform `all_gather` and everything works just fine.

### Potential Solution

1. Upgrade the PyTorch dependency from 2.4.0 to >=2.5.0, for uneven sized tensors are supported in PyTorch>=2.5.0 (https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather). However, I haven't been able to try this out because the PyTorch>=2.5.0 is only compatible with vLLM>=0.6.4, whereas the current veRL framework supports vLLM<=0.6.3.
2. Set `actor_rollout_ref.rollout.tensor_model_parallel_size = 1`.
3. Ensure `data.train_batch_size` and `data.val_batch_size` are divisible by the world size.

I think either solution could solve the problem, but the latter 2 are just workarounds. I'm wondering maybe the developers could help inspect the first solution so that FSDP-vLLM Rollout is more robust to different configurations.

**Comment by PeterSH6** [comment]
Hi @nwiad, thanks for your detailed investigation and suggestion.

Yeah, the current implementation requires users to specify the `data.train_batch_size` and `data.val_batch_size` correctly w.r.t the parallel size (and also world_size in FSDP). This is because our DataProto chunk method can deal with the case where `batch size % world_size != 0` while the all_gather cannot. In fact, users should make sure that both `train_batch_size` and  `val_batch_size`  can be evenly divided by the `world_size` in the FSDP case. We may add some assertions in advance.

The solution 2 you proposed may not be a good workaround as it may cause OOM in larger models

For solution 1, this could be a good solution for validation. But for training, this may cause some silent internal issues in all-gather or all-reduce. I'll take some time to check if this feature is feasible in our case before upgrading Torch >= 2.5

**Comment by nwiad** [comment]
> Hi @nwiad, thanks for your detailed investigation and suggestion.
> 
> Yeah, the current implementation requires users to specify the `data.train_batch_size` and `data.val_batch_size` correctly w.r.t the parallel size (and also world_size in FSDP). This is because our DataProto chunk method can deal with the case where `batch size % world_size != 0` while the all_gather cannot. In fact, users should make sure that both `train_batch_size` and `val_batch_size` can be evenly divided by the `world_size` in the FSDP case. We may add some assertions in advance.
> 
> The solution 2 you proposed may not be a good workaround as it may cause OOM in larger models
> 
> For solution 1, this could be a good solution for validation. But for training, this may cause some silent internal issues in all-gather or all-reduce. I'll take some time to check if this feature is feasible in our case before upgrading Torch >= 2.5

Absolutely. I think it's a good idea to simply add some assertions which could prevent users from triggering this problem at minimal effort.

Admittedly, for users who are familiar with MPI programming, it's a common sense that tensors should be identical in size before performing `all_gather`. It's just that users may not be aware of the communication details of this framework or rather, some are completely new to collective communication. So I think maybe some precautions in the docs/tutorial are helpful for making this framework more inclusive.

BTW, I read the codes thoroughly. It's probably the most concise and understandable RLHF framework I've seen. Thanks to its fair extensibility, I managed to implement the GRPO algorithm by extending `RayPPOTrainer`.

👍 Nice work

**Comment by PeterSH6** [comment]
@nwiad looking forward to your GRPO contribution!

In fact, we have also left some APIs for GRPO implementation, such as setting `use_critic=False`.
We also managed to make some refactors for better extensibility recently (probably today):
- Move actor/critic/hybrid_engine/reward_model/rollout/workers out of verl/trainer/ppo directory to verl/trainer directory.
These packages are designed for reusing between different RL algorithms.

Therefore, it would be better if you could implement GRPO only by modifying `core_algos.py` and `ray_trainer.py`.
For the n-times generation feature, you could modify the `n` parameter in vLLM's sampling_param to a value greater than 1. However, this may require reconstructing the n-responses alongside their corresponding prompts in `vllm_rollout.py`.

Hope this will help you and feel free to ask any questions during your development.

**Comment by nwiad** [comment]
> @nwiad looking forward to your GRPO contribution!
> 
> In fact, we have also left some APIs for GRPO implementation, such as setting `use_critic=False`. We also managed to make some refactors for better extensibility recently (probably today):
> 
> * Move actor/critic/hybrid_engine/reward_model/rollout/workers out of verl/trainer/ppo directory to verl/trainer directory.
>   These packages are designed for reusing between different RL algorithms.
> 
> Therefore, it would be better if you could implement GRPO only by modifying `core_algos.py` and `ray_trainer.py`. For the n-times generation feature, you could modify the `n` parameter in vLLM's sampling_param to a value greater than 1. However, this may require reconstructing the n-responses alongside their corresponding prompts in `vllm_rollout.py`.
> 
> Hope this will help you and feel free to ask any questions during your development.

Yes I noticed the preserved API `use_critic` in the code, which I presume is intended for future GRPO/ReMax extensions. So I follow this intention to disable critic when using GRPO.

As for group-relative sampling (i.e. n-times generation), I found it error-prone to reconstruct common modules like `vllm_rollout`. So I decided to hack the `collate_fn` param of `train_dataloader`, where I repeat each sample for `group_size` times, making the actual training batch `group_size` times larger. This saves the effort of having to deal with inconsistent sized batch before/after rollout. Meanwhile, it's necessary to hack `ppo_mini_batch_size` by multiplying `group_size` as well. This is because rollout results of the same group must be used to optimize the policy in the same optimization step. Such adaptation is invisible to the original workflow and makes it easy for debugging.

A concern of this adaptation is that copying prompts for `group_size` times isn't cost-efficient enough. I think you are right about using `n` param in vLLM's sampling param. This may complicate the workflow a bit but will save us memory.

I will try to optimize this part real quick. Thank you for your advice.
